### PR TITLE
Fix #5976: Don't assume `T <:< (=>T)`.

### LINKED ISSUE #5976: Eta-expansion for method with by-name parameter fail at runtime
```scala
def f(i: => Int) = i + i
implicit def ups(f: ((=>Int) => Int)): (Int => Int) = x => f(x)
val res = List(42).map(f)
```
work with scalac but fail with dotty even with -language:Scala2
Example come from : https://github.com/scala/bug/issues/11414

~~~stacktrace
java.lang.ClassCastException: class java.lang.Integer cannot be cast to class scala.Function0 (java.lang.Integer is in module java.base of loader 'bootstrap'; scala.Function0 is in unnamed module of loader 'app')
	at scala.collection.immutable.List.map(List.scala:286)
	at rs$line$1$.<init>(rs$line$1:3)
	at rs$line$1$.<clinit>(rs$line$1)
	at rs$line$1.res(rs$line$1)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at dotty.tools.repl.Rendering.$anonfun$3(Rendering.scala:57)
	at scala.Option.map(Option.scala:163)
	at dotty.tools.repl.Rendering.valueOf(Rendering.scala:57)
	at dotty.tools.repl.Rendering.renderVal(Rendering.scala:80)
	at dotty.tools.repl.ReplDriver.displayMembers$3$$anonfun$3(ReplDriver.scala:274)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at scala.collection.TraversableLike.map(TraversableLike.scala:237)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at dotty.tools.repl.ReplDriver.displayMembers$7(ReplDriver.scala:274)
	at dotty.tools.repl.ReplDriver.displayDefinitions$$anonfun$3$$anonfun$2(ReplDriver.scala:301)
	at scala.Option.map(Option.scala:163)
	at dotty.tools.repl.ReplDriver.displayDefinitions$$anonfun$1(ReplDriver.scala:301)
	at dotty.tools.dotc.core.Periods.atPhase(Periods.scala:25)
	at dotty.tools.dotc.core.Phases.atPhase(Phases.scala:35)
	at dotty.tools.repl.ReplDriver.displayDefinitions(ReplDriver.scala:306)
	at dotty.tools.repl.ReplDriver.compile$$anonfun$2(ReplDriver.scala:230)
	at scala.util.Either.fold(Either.scala:191)
	at dotty.tools.repl.ReplDriver.compile(ReplDriver.scala:230)
	at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:182)
	at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:127)
	at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:130)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput$$anonfun$2$$anonfun$1(ReplDriver.scala:143)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at scala.Console$.withErr(Console.scala:196)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput$$anonfun$1(ReplDriver.scala:143)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at scala.Console$.withOut(Console.scala:167)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:143)
	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:130)
	at dotty.tools.repl.Main$.main(Main.scala:6)
	at dotty.tools.repl.Main.main(Main.scala)
~~~

### PR DESCRIPTION
**Pull request by odersky**
It's incorrect to assume `T <:< (=> T)` since, taken as parameters,
the erasure of the two types is different: `|T|` vs `Function1`.

To compenate, we need to `widenExpr` in various places.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 1]
```diff
@@ -603,7 +603,7 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
    * argument trees.
    */
   class ApplicableToTreesDirectly(methRef: TermRef, targs: List[Type], args: List[Tree], resultType: Type)(implicit ctx: Context) extends ApplicableToTrees(methRef, targs, args, resultType)(ctx) {
-    override def argOK(arg: TypedArg, formal: Type): Boolean = argType(arg, formal) <:< formal
+    override def argOK(arg: TypedArg, formal: Type): Boolean = argType(arg, formal) <:< formal.widenExpr
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala]: Do we have some guarantee that the result of `argType` cannot be an ExprType here ? Even if we do, it seems more regular to widen both sides.

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 2]
```diff
@@ -98,7 +98,7 @@ object Implicits {
             else if (mt.paramInfos.lengthCompare(1) == 0 && {
                   var formal = widenSingleton(mt.paramInfos.head)
                   if (approx) formal = wildApprox(formal)
-                  ctx.test(implicit ctx => argType relaxed_<:< formal)
+                  ctx.test(implicit ctx => argType relaxed_<:< formal.widenExpr)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Implicits.scala]: Same question here.

**Comment by smarter** [comment]
I'm wondering if we should introduce a `widen_<:<` to make the pattern more formal.

File: compiler/src/dotty/tools/dotc/typer/Applications.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Applications.scala]: The case we dropped in TypeComparer is where the RHS is an ExprType. `=> T` was never a subtype of `T`.

**Comment by odersky** [comment]
> I'm wondering if we should introduce a `widen_<:<` to make the pattern more formal.

I had a look at it. We have 8 occurrences of the pattern `<:< t.widenExpr` but 5 use `<:<` and 3 use `frozen_<:<`. So we'd have to create two new methods to capture all these cases. My tendency is to not do that, though.

File: compiler/src/dotty/tools/dotc/typer/Applications.scala (hunk 1 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/typer/Applications.scala]: Right, but if both `argType` and `formal` are `=> T`, then `argOK` used to return true but will return false now that we widen `formal`.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Applications.scala]: I see. I believe argTypes are never ExprTypes. They might be TermRefs with an underlying ExprType. But we skip the ExprType when we do a widen on these. `lub` also does `widen`, not `widenExpr`. So I think we are good.

**Pull request merged**